### PR TITLE
Ko 229

### LINKED ISSUE #229: track how many pages are in a table
## Summary

- we have pagination setup, but currently have no way of checking how many pages are in a table

### Basic example

- 25 users are added to abacus (equal to a page), the number of pages when paginating should be incremented; but currently isn't

### Motivation

- if 25 users are added or deleted the number of pages will be inaccurate

### PR DESCRIPTION
**Pull request by MrKevinOConnell**
# Description

- when retrieving a page from the DB, there is now a check to see how many pages are in the table, and if the given page does not exist.

Screen Shot 2021-11-15 at 2 34 04 PM
Screen Shot 2021-11-15 at 2 34 13 PM
Screen Shot 2021-11-15 at 2 34 36 PM

Fixes #229 

- goes off https://github.com/acm-mu/abacus/pull/228

## Type of change

- [x] 💡 New feature

  
# How Has This Been Tested?

- upon loading the users table the correct number of pages should show (have 5 users per page at the moment for testing reasons 

- [x] Tested individually.

# Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [x] My changes do not break any features.

### CODE REVIEW

File: frontend/src/pages/admin/Clarifications.tsx [hunk 1]
```diff
@@ -86,25 +117,24 @@ const Clarifications = (): JSX.Element => {
...
   useEffect(() => {
-    loadClarifications()
+    loadClarifications(page)
     return () => {
       setMounted(false)
     }
-  }, [])
+  }, [page])
 
+  // <ClarificationModal trigger={<Button content="Create Clarification" />} callback={loadClarifications} />
```
Comment (a1g3) [inline frontend/src/pages/admin/Clarifications.tsx:130]: Can remove this if its not needed.

**Review by a1g3 — CHANGES_REQUESTED** [review]
What is the relation between this PR and PR #228?

File: frontend/src/pages/admin/Clarification.tsx [hunk 2]
```diff
@@ -32,7 +32,7 @@ const ClarificationPage = (): JSX.Element => {
         Authorization: `Bearer ${localStorage.accessToken}`
       }
     })
-    if (response.ok && isMounted) {
+    if (response.ok) {
```
[the remark this answers is not in the document]
Comment (MrKevinOConnell) [reply frontend/src/pages/admin/Clarification.tsx:35]: After a couple calls are made to update the page the component stops being mounted, so I just removed it (not sure when actions would occur on a non used component anyways)

File: backend/src/api/auth/getTableSize.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    return
+  }
+  const params = matchedData(req)
+
+  try {
+    //makes call to db to get table size.
+    const tableSize = await contest.get_table_size(tableName, params)
+    //divides by the number of items per page.
+
+    const pages = tableSize ? tableSize / 5 : 0
```
[the remark this answers is not in the document]
Comment (MrKevinOConnell) [reply backend/src/api/auth/getTableSize.ts:64]: tablesize returns the total number of items in a table. For the purpose of showing the pagination functionality I set each page to 5 items, what that does is divide by 5 to get the number of pages

File: backend/src/services/db/dynamo.ts [hunk 4]
```diff
@@ -11,11 +11,51 @@ export default class DynamoDB extends Database {
...
+    return new Promise(async (resolve, reject) => {
+      // might change exclusivestartkey below, so I disable prefer const
+      // eslint-disable-next-line prefer-const
+      let params: ScanInput = { TableName }
+      if (query) {
+        if (query.args) {
+          const entries = Object.entries(query.args)
+          if (entries.length > 0) {
+            params.FilterExpression = entries.map((e) => `#${e[0]} = :${e[0]}`).join(' AND ')
+            params.ExpressionAttributeNames = Object.assign({}, ...entries.map((x) => ({ [`#${x[0]}`]: x[0] })))
```
[the remark this answers is not in the document]
Comment (MrKevinOConnell) [reply backend/src/services/db/dynamo.ts:24]: lmao, yea that's jack

File: backend/src/api/auth/getTableSize.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  },
+  username: {
+    in: ['body', 'query'],
+    isString: true,
+    optional: true,
+    errorMessage: 'String username is not supplied'
+  }
+}
+
+export const getTableSize = async (req: Request, res: Response) => {
```
[the remark this answers is not in the document]
Comment (MrKevinOConnell) [reply backend/src/api/auth/getTableSize.ts:50]: this makes it so every time we make a call to get a page, we can
1. make sure the page is still valid
2. update the number of pages for future calls

what do you suggest is a better way of doing this?

File: frontend/Dockerfile [hunk 6]
```diff
@@ -1,5 +1,5 @@
 # pull official base image
-FROM node
+FROM node:16
```
[the remark this answers is not in the document]
Comment (MrKevinOConnell) [reply frontend/Dockerfile:2]: Yea, I figured this out here https://github.com/acm-mu/abacus/pull/228, just reverted to 16

File: backend/src/services/db/mongo.ts [hunk 7]
```diff
@@ -22,12 +22,16 @@ export default class MongoDB extends Database {
     })
   }
 
-  scan(TableName: string, query?: ScanOptions): Promise<Item[]> {
+  scan(TableName: string, query?: ScanOptions, page?: number): Promise<Item[]> {
```
[the remark this answers is not in the document]
Comment (MrKevinOConnell) [reply backend/src/services/db/mongo.ts:25]: to be honest, I tried implementing it into args of ScanOptions but had some trouble due to the key being checked into mongo as part of the scan call (not sure how to delete the record from the property.)

File: frontend/src/pages/admin/Clarifications.tsx [hunk 8]
```diff
@@ -25,31 +25,59 @@ const Clarifications = (): JSX.Element => {
...
+          Authorization: `Bearer ${localStorage.accessToken}`,
+          'Content-Type': 'application/json'
+        }
+      })
+
+      const numberOfPages = await tableSizeRes.json()
+      const { tableSize } = numberOfPages
+
+      setNumberOfPages(Math.ceil(tableSize))
+      if (tableSize < numberOfPages) {
```
[the remark this answers is not in the document]
Comment (MrKevinOConnell) [reply frontend/src/pages/admin/Clarifications.tsx:54]: this is if the call comes back and tablesize is higher than numberpages (due to new clarifications being made

File: backend/src/services/db/dynamo.ts [hunk 9]
```diff
@@ -11,11 +11,51 @@ export default class DynamoDB extends Database {
     this.db = new DocumentClient()
   }
 
-  scan(TableName: string, query: ScanOptions): Promise<Item[]> {
+  count(TableName: string, query: ScanOptions): Promise<number> {
```
[the remark this answers is not in the document]
Comment (MrKevinOConnell) [reply backend/src/services/db/dynamo.ts:14]: it doesn't but we need a function for each table

File: backend/src/api/submissions/getSubmissions.ts [hunk 10]
```diff
@@ -178,21 +181,22 @@ export const getSubmissions = async (req: Request, res: Response): Promise<void>
...
-          submission.status = 'pending'
-          submission.score = 0
-          submission.tests = submission.tests?.map((test: Test) => ({ ...test, result: '' }))
-        }
-        return submission
-      })
-      .filter((submission) => showToUser(req.user, submission.problem, settings))
-
-    res.send(transpose(submissions, 'sid'))
+    if (submissions !== []) {
```
[the remark this answers is not in the document]
Comment (MrKevinOConnell) [reply backend/src/api/submissions/getSubmissions.ts:187]: I was getting errors in the case of there being no users in a table, so i added a couple checks

File: frontend/src/pages/admin/Clarifications.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (MrKevinOConnell) [reply frontend/src/pages/admin/Clarifications.tsx:52]: > What is the relation between this PR and PR #228?

this goes off of it, just adds the implementation of pages

File: frontend/src/pages/admin/Clarification.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jl5her) [reply frontend/src/pages/admin/Clarification.tsx:35]: If the user clicks away from the page before the server responds, this check will prevent react from updating the page that is no longer mounted.

**Pull request merged**